test/e2e/e2e_test.go

```diff
@@ -70,11 +79,79 @@ func TestMain(m *testing.M) {
 		cancel()
 	}()
 
+	if globalOpts.ArtifactDir != "" {
+		go setupMetricsEndpoint(testContext, log)
+		go e2eObserverControllers(testContext, log, globalOpts.ArtifactDir)
+		defer dumpTestMetrics(log, globalOpts.ArtifactDir)
```

Comment:
Can't use a `defer` in `TestMain` because it must call `os.Exit()` which won't honor `defer` statements.

Reply:
But the context never gets cancelled during a normal run without failure, so that would mean we can't dump them there. If I understand the go doc correctly, there is no need to call `os.Exit` in TestMain:

```
TestMain runs in the main goroutine and can do
whatever setup and teardown is necessary around a call to m.Run. m.Run will
return an exit code that may be passed to os.Exit. If TestMain returns, the
test wrapper will pass the result of m.Run to os.Exit itself.
```

Reply:
>If TestMain returns, the test wrapper will pass the result of m.Run to os.Exit itself.

Never noticed that detail, thank you! So we can just remove all calls to `os.Exit` and `m.Run()` entirely from `TestMain`... I just tried this in a local test and it works just like you said. Thank you!

Reply:
>But the context never gets cancelled during a normal run without failure

That seems like an oversight/bug, shouldn't `TestMain` also be deferring a cancellation of the root context?

Reply:
> That seems like an oversight/bug, shouldn't TestMain also be deferring a cancellation of the root context?

True, added that. And realized that the pattern of not calling `t.Run()` won't give us what we want: Execute the `defer` statements _after_ all tests were executed, it would just execute it before them. So I instead updated this to follow the suggestion of factoring out the `t.Run()` call into a new function and `defer` there